File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/course/CourseRepo.java
package kh.farrukh.sortmapperexamplejava.course;

import java.util.List;
import java.util.Optional;
import kh.farrukh.sortmapperexamplejava.course.model.CourseShortInfoResponseDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

@Repository
public interface CourseRepo extends JpaRepository<CourseEntity, Long> {

  Optional<CourseEntity> findByIdAndDeletedAtIsNull(Long id);

  Page<CourseEntity> findAllByDeletedAtIsNull(Pageable pageable);

  @Query("""
      select new kh.farrukh.sortmapperexamplejava.course.model.CourseShortInfoResponseDTO(
        course.id,
        course.title,
        course.teacher.firstName
      )
      from CourseEntity course
      where course.deletedAt is null
      """)
  List<CourseShortInfoResponseDTO> findAllShortInfo(Sort sort);

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/course/CourseRepo.kt
package kh.farrukh.sortmapperexample.course

import kh.farrukh.sortmapperexample.course.model.CourseShortInfoResponseDTO
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

@Repository
interface CourseRepo : JpaRepository<CourseEntity, Long> {

    fun findByIdAndDeletedAtIsNull(id: Long): CourseEntity?

    fun findAllByDeletedAtIsNull(pageable: Pageable): Page<CourseEntity>

    @Query(
        """
           select new kh.farrukh.sortmapperexample.course.model.CourseShortInfoResponseDTO(
                course.id,
                course.title,
                course.teacher.firstName
           )
           from CourseEntity course
           where course.deletedAt is null
        """
    )
    fun findAllShortInfo(sort: Sort): List<CourseShortInfoResponseDTO>

}